private/buf/bufconvert/bufconvert.go

```diff
@@ -0,0 +1,151 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufconvert
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/bufbuild/buf/private/pkg/stringutil"
+	"go.opencensus.io/trace"
+)
+
+const (
+	// MessageEncodingBin is the binary image encoding.
+	MessageEncodingBin MessageEncoding = iota + 1
+	// MessageEncodingJSON is the JSON image encoding.
+	MessageEncodingJSON
+	// formatBin is the binary format.
+	formatBin = "bin"
+	// formatJSON is the JSON format.
+	formatJSON = "json"
+)
+
+var (
+	// MessageEncodingFormatsString is the string representation of all message encoding formats.
+	//
+	// This does not include deprecated formats.
+	MessageEncodingFormatsString = stringutil.SliceToString(messageEncodingFormats)
+	// sorted
+	messageEncodingFormats = []string{
+		formatBin,
+		formatJSON,
+	}
+)
+
+// MessageEncoding is the encoding of the message
+type MessageEncoding int
+
+type MessageEncodingRef interface {
+	Path() string
+	MessageEncoding() MessageEncoding
+}
+
+func NewMessageEncodingRef(
+	ctx context.Context,
+	value string,
+	defaultEncoding MessageEncoding,
+) (MessageEncodingRef, error) {
+	ctx, span := trace.StartSpan(ctx, "new_message_encoding_ref")
+	defer span.End()
+	path, messageEncoding, err := getPathAndMessageEncoding(ctx, value, defaultEncoding)
+	if err != nil {
+		return nil, err
+	}
+	return newMessageEncodingRef(path, messageEncoding), nil
+}
+
+func getPathAndMessageEncoding(
+	ctx context.Context,
+	value string,
+	defaultEncoding MessageEncoding,
+) (string, MessageEncoding, error) {
+	path, options, err := getRawPathAndOptions(value)
+	if err != nil {
+		return "", 0, err
+	}
+	messageEncoding := parseMessageEncodingExt(filepath.Ext(path), defaultEncoding)
+	if format, ok := options["format"]; ok {
+		messageEncoding, err = parseMessageEncodingFormat(format)
+		if err != nil {
+			return "", 0, err
+		}
+	}
+	return path, messageEncoding, nil
+}
+
+func parseMessageEncodingExt(ext string, defaultEncoding MessageEncoding) MessageEncoding {
+	switch strings.TrimPrefix(ext, ".") {
+	case formatBin:
+		return MessageEncodingBin
+	case formatJSON:
+		return MessageEncodingJSON
+	default:
+		return defaultEncoding
+	}
+}
+
+func parseMessageEncodingFormat(format string) (MessageEncoding, error) {
+	switch format {
+	case formatBin:
+		return MessageEncodingBin, nil
+	case formatJSON:
+		return MessageEncodingJSON, nil
+	default:
+		return 0, fmt.Errorf("invalid format for message: %q", format)
+	}
+}
+
+// rawPath will be non-empty
+func getRawPathAndOptions(value string) (string, map[string]string, error) {
```

Comment:
It'd be great if we could share this logic between `buffetch` and this - copying complex code like this can cause problems.

In both cases we're working with some sort of reference - maybe we could move this to `bufref.GetRawPathAndOptions` and have both `bufconvert` and `buffetch` import from it? I'm not sure what other package names to use here (cc @bufdev).

Reply:
Idk I'd just choose a package name, doesn't need to make too much sense